### PR TITLE
Introduce WithEnvironment overload for custom connection string keys #3002

### PR DESCRIPTION
**Pull request by danikishin**
This PR introduces a new overload for `WithEnvironment` method in the `ResourceBuilderExtensions` class, which allows setting the connection string of a resource under a custom environment variable name.

#### Changes
- Added `WithEnvironment` overload to accept a custom environment variable name.

#### Background
The issue #3002 describes a scenario where existing projects do not use the 'ConnectionStrings__' prefix for environment variables. 

#### Example Usage
```csharp
builder.AddProject<Projects.Web>("web").WithEnvironment("CustomEnvVarName", redis);
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3239)

### CODE REVIEW

**Comment by danikishin** [comment]
> @danikishin please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement



> ```
> @dotnet-policy-service agree
> ```

@dotnet-policy-service agree

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -115,6 +115,33 @@ public static IResourceBuilder<T> WithEnvironment<T>(this IResourceBuilder<T> bu
...
+        return builder.WithEnvironment(async context =>
+        {
+            var connectionString = await sourceBuilder.Resource.GetConnectionStringAsync().ConfigureAwait(true);
+
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new DistributedApplicationException($"A connection string for '{sourceBuilder.Resource.Name}' could not be retrieved.");
+            }
+
+            context.EnvironmentVariables[envVarName] = connectionString;
```
Comment (davidfowl) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:141]:
```suggestion
            context.EnvironmentVariables[envVarName] = new ConnectionStringReference(sourceBuilder.Resource, optional: false);
```

**Comment by davidfowl** [comment]
This should have tests. Look at the existing WithEnvironment tests

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 2]
```diff
@@ -115,6 +115,26 @@ public static IResourceBuilder<T> WithEnvironment<T>(this IResourceBuilder<T> bu
...
+    /// <param name="envVarName">The name of the environment variable under which the connection string will be set.</param>
+    /// <param name="sourceBuilder">The resource builder of the referenced service from which to pull the connection string.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> WithEnvironment<T>(
+        this IResourceBuilder<T> builder,
+        string envVarName,
+        IResourceBuilder<IResourceWithConnectionString> sourceBuilder)
+        where T : IResourceWithEnvironment
+    {
+        return builder.WithEnvironment(async context =>
```
Comment (davidfowl) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:132]:
```suggestion
        return builder.WithEnvironment(context =>
```

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 3]
```diff
@@ -115,6 +115,26 @@ public static IResourceBuilder<T> WithEnvironment<T>(this IResourceBuilder<T> bu
...
+    /// </summary>
+    /// <typeparam name="T">The destination resource type.</typeparam>
+    /// <param name="builder">The destination resource builder to which the environment variable will be added.</param>
+    /// <param name="envVarName">The name of the environment variable under which the connection string will be set.</param>
+    /// <param name="sourceBuilder">The resource builder of the referenced service from which to pull the connection string.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> WithEnvironment<T>(
+        this IResourceBuilder<T> builder,
+        string envVarName,
+        IResourceBuilder<IResourceWithConnectionString> sourceBuilder)
```
Comment (davidfowl) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:129]: Do we call it sourceBuilder anywhere else?

File: tests/Aspire.Hosting.Tests/WithEnvironmentTests.cs [hunk 4]
```diff
@@ -171,6 +171,28 @@ public async Task EnvironmentVariableExpressions()
...
+        var targetBuilder = testProgram.AppBuilder.AddContainer("targetContainer", "targetImage");
+
+        string envVarName = "CUSTOM_CONNECTION_STRING";
+
+        // Act
+        targetBuilder.WithEnvironment(envVarName, sourceBuilder);
+        testProgram.Build();
+
+        // Call environment variable callbacks with the Publish operation.
+        var config = await EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync(targetBuilder.Resource, DistributedApplicationOperation.Publish);
```
Comment (davidfowl) [inline tests/Aspire.Hosting.Tests/WithEnvironmentTests.cs:190]: Can you also test run mode? (DistributedApplicationOperation.Run)

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 5]
```diff
@@ -115,6 +115,26 @@ public static IResourceBuilder<T> WithEnvironment<T>(this IResourceBuilder<T> bu
...
     }
 
+    /// <summary>
+    /// Configures a connection string from a referenced resource builder to be set as an environment variable with a custom name.
+    /// </summary>
+    /// <typeparam name="T">The destination resource type.</typeparam>
+    /// <param name="builder">The destination resource builder to which the environment variable will be added.</param>
+    /// <param name="envVarName">The name of the environment variable under which the connection string will be set.</param>
+    /// <param name="resource">The resource builder of the referenced service from which to pull the connection string.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
```
Comment (davidfowl) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:125]: Can you make sure the doc comments are consistent with the other overloads?

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 6]
```diff
@@ -115,6 +115,26 @@ public static IResourceBuilder<T> WithEnvironment<T>(this IResourceBuilder<T> bu
         });
     }
 
+    /// <summary>
+    /// Configures a connection string from a referenced resource builder to be set as an environment variable with a custom name.
```
Comment (davidfowl) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:119]:
```suggestion
    /// Adds an environment variable to the resource with the connection string from the referenced resource.
```

**Review by davidfowl — APPROVED** [review]

**Pull request merged**